mlflow/tensorflow.py

```diff
@@ -333,3 +343,125 @@ def predict(self, df):
             raw_preds = self.tf_sess.run(self.output_tensors, feed_dict=feed_dict)
             pred_dict = {column_name: values.ravel() for column_name, values in raw_preds.items()}
             return pandas.DataFrame(data=pred_dict)
+
+
+class __MLflowTfKerasCallback(Callback):
+    def __init__(self):
+        if mlflow.active_run() is None:
+            mlflow.start_run()
+
+    def __enter__(self):
+        pass
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        pass
+
+    def on_epoch_end(self, epoch, logs=None):
+        pass
+
+    def on_train_end(self, logs=None):
+        mlflow.log_param('learning_rate', tensorflow.keras.backend.eval(self.model.optimizer.optimizer._lr))
+        mlflow.log_param('epsilon', tensorflow.keras.backend.eval(self.model.optimizer.optimizer._epsilon))
+        mlflow.log_param('optimizer_name', self.model.optimizer.optimizer._name)
+        #TODO: Fix for keras log_model not saving TF optimizers
+        mlflow.keras.log_model(self.model, None, None)
+
+_metric_queue = []
+
+_thread_pool = concurrent.futures.ThreadPoolExecutor(max_workers=1)
+
+
+def flush_queue():
+    global _metric_queue
+    try:
+        client = mlflow.tracking.MlflowClient()
+        if mlflow.active_run() is not None:
+            client.log_batch(mlflow.active_run().info.run_id, metrics=_metric_queue, params=[], tags=[])
+    except Exception as e:
+        warnings.warn("Logging to MLflow failed: " + str(e))
+    finally:
+        _metric_queue = []
+
+
+atexit.register(flush_queue)
+
+
+def add_to_queue(key, value, step):
+    met = Metric(key=key, value=value, timestamp=int(time.time()*1000), step=step)
+    _metric_queue.append(met)
+    if len(_metric_queue) >= 500:
+        flush_queue()
+
+
+def _log_event(event):
+    if mlflow.active_run() is None:
+        mlflow.start_run()
+    if event.WhichOneof('what') == 'summary':
+        summary = event.summary
+        for v in summary.value:
+            if v.HasField('simple_value'):
+                if event.step % 100 == 0:
+                    _thread_pool.submit(add_to_queue, key=v.tag, value=v.simple_value, step=event.step)
+
+
+def _checkpoint_model(model_path):
+    mlflow.log_artifact(model_path)
+
+
+def autolog():
+    import tensorflow
+    from tensorflow.python.summary.writer.event_file_writer import EventFileWriter
+    from tensorflow.python.summary.writer.event_file_writer_v2 import EventFileWriterV2
+
+    def contains_tensorboard_callback(lst):
+        for x in lst:
+            if isinstance(x, tensorflow.keras.callbacks.TensorBoard):
+                return True
+
+    @gorilla.patch(tensorflow.estimator.Estimator)
+    def export_saved_model(self, *args, **kwargs):
```

Comment:
We should consider adding an identical patch for `export_savedmodel`, which is available in older TF versions (while `export_saved_model` is not), e.g: https://www.tensorflow.org/versions/r1.12/api_docs/python/tf/estimator/Estimator#export_savedmodel

Reply:
Done, thanks!